packages/language-server/src/plugins/typescript/LSAndTSDocResolver.ts

```diff
@@ -66,7 +66,13 @@ export class LSAndTSDocResolver {
         // where multiple files and their dependencies
         // being loaded in a short period of times
         docManager.on('documentOpen', (document) => {
+            const path = document.getFilePath();
+            const isNewFile = path && !this.globalSnapshotsManager.get(path);
             handleDocumentChange(document);
+            if (isNewFile && document.openedByClient) {
```

Comment:
Does this mean we would kick off the costly "update project files" logic on _every_ new file created through the editor UI? Is there a way to do this in less cases? "update project files" reads the file system, creates a new program version etc, which seems costly.

Reply:
I changed to detection to file watch since #2393 also needs it. I also delay the check to the next time the language service is used. TypeScript also has some directory watching info in the config file parse result. So we can skip the project file check if the file isn't in the target directories. This info is also used to watch directories outside the root using the fallback watch or when the client supports it. This is added because the optimization check might break a way to workaround directory watching outside the workspace, not sure if anyone actually uses it though 😅.